### ISSUE TITLE
AV stack overflow with Intel TBB malloc

### ISSUE THREAD

**Issue by robertodellapasqua** [winddriver/Delphi-Cross-Socket#39]
please check the problems of this library with intel tbb malloc (causing AV stack overflow)

https://en.delphipraxis.net/topic/636-heavy-bug-102-103-rio-passing-parameter-pointers-in-anonymous-methods/

do we have memory overrun or use-after-free in the code (something as out of order getmem/freemem)?

**Comment by robertodellapasqua** [comment]
please do a QA cycle over the code
you should use madExcept, Nexus SleuthQA, TotalQA or EurekaLog to trap memory leaks, buffer overrun, api fails
thank you

**Comment by robertodellapasqua** [comment]
so, the author, Souledge, if you have spare time try to download my tbbmalloc wrapper (or any other else) https://en.delphipraxis.net/applications/core/interface/file/attachment.php?id=580 then see the AV happening on start, and failing I/O read write sockets with high rate. Somewhere, something of wrong should be.
I can help if you wish. 
Best.

**Comment by winddriver** [comment]
Thank for your attention and patience. With Delphi's default memory manager, I tested the memory leakage of my library with madExcept and found no leakage. I haven't used Intel TBB yet. I'll take the time to test it.

**Comment by pony5551** [comment]
我用tbb和msvcrt.dll里的好像都有泄漏！
另外还有最新版的fastmm好像也是有泄漏，我从github上clone的fastmm
作者可以用新版fastmm试试！

**Comment by robertodellapasqua** [comment]
這不是洩漏問題，而是死鎖競賽循環

**Comment by robertodellapasqua** [comment]
IMHO is the compiler producing code artifacts on encapsulation of abstract/virtual classes/interfaces, let's see what Souledge will discover

**Comment by robertodellapasqua** [comment]
sorry I did a wrong information, and my english is not good, sorry
I want mean that "dynamic" vs "virtual" methods when you use abstraction, between the two prefer the virtual (VMT) instead of dynamic (DMT)

Virtual versus Dynamic
In Delphi for Win32, virtual and dynamic methods are semantically equivalent. However, they differ in the implementation of method-call dispatching at run time: virtual methods optimize for speed, while dynamic methods optimize for code size.

In general, virtual methods are the most efficient way to implement polymorphic behavior. Dynamic methods are useful when a base class declares many overridable methods that are inherited by many descendent classes in an application, but only occasionally overridden.

Note: Only use dynamic methods if there is a clear, observable benefit. Generally, use virtual methods.

**Comment by pony5551** [comment]
Thank you very much for your answer! I see what you mean!

But what is the reason for TBB's av depends on what souledge does!

**Comment by pony5551** [comment]
我怀疑是delphi的类型问题，导致这些三方内存管理器不认识一些类型无法正确释放，
fastmm4新版本可能是自身的问题与delphi cross socket无关

**Comment by robertodellapasqua** [comment]
> 另外用google的tcmm测试了一下，发现也抛出异常
> E:\gperftools\gperftools-master2.7\src\tcmalloc.cc:332] Attempt to free invalid pointer 0x356c5f0
> E:\gperftools\gperftools-master2.7\src\tcmalloc.cc:332] Attempt to free invalid pointer 0x356bba0

when you get these errors?
can you paste stack trace?
anyway I wish that Souledge will found the trouble doing a workaround, because Intel TBB it's absolutely the most scalable mem manager for server apps

**Comment by pony5551** [comment]
intelTBB报错如下:

Windows 10 (Version 10.0, Build 17763, 64-bit Edition)
TIocpCrossSocket._NewAccept.AcceptEx : System Error.  Code: 0(0000), 操作成功完成。
CrossHttpServer start at port:8000, IO threads:1
Press enter stop
TCrossHttpServer Io线程ID 9288, 异常 EAccessViolation, Access violation at address 000000006BA85374 in module 'tbbmalloc_x64.dll'. Write of address 0000000000000010
TCrossHttpServer Io线程ID 9288, 异常 EAccessViolation, Access violation at address 0000000000417469 in module 'CrossHttpConsole.exe'. Write of address 0000000000007236

我只开了一个线程

**Comment by pony5551** [comment]
function TIocpCrossSocket._NewIoData: PPerIoData;
begin
  GetMem(Result,SizeOf(TPerIoData));
  //System.New(Result);
  FillChar(Result^, SizeOf(TPerIoData), 0);
end;

procedure TIocpCrossSocket._FreeIoData(P: PPerIoData);
begin
  P.CrossData := nil;
  //System.Dispose(P);
  FreeMem(P, SizeOf(TPerIoData));
end;

如果代码改成这样，程序退出时会卡住？

**Comment by pony5551** [comment]
一个线程！fastmm弹出
---------------------------
CrossHttpConsole.exe: MM Operation after uninstall.
---------------------------
FastMM has detected a FreeMem call after FastMM was uninstalled.
---------------------------
确定   
---------------------------

**Comment by pony5551** [comment]
另外10.3编译32位时，(?<!\.)\*这个表达式有问题，64位正常！
这算是delphi的bug吧！

**Comment by winddriver** [comment]
测试了一下最新的FastMM4，的确有一些问题，不过也有一些新的发现

在将 System.New 和System.Dispose 替换成 GetMem 和 FreeMem 后结合 FastMM4 测试，FastMM4 需要调整一下编译开关，在 FastMM4Options.inc 中开启 AssumeMultiThreaded 和 NeverUninstall
![image](https://user-images.githubusercontent.com/3221597/52171914-57a20f80-27a0-11e9-8b5d-01816f517b99.png)
![image](https://user-images.githubusercontent.com/3221597/52171916-64266800-27a0-11e9-96de-4cbdb94176a6.png)


测试结果如下：

- 2000并发，无论跑多少次多少请求，都不会有任何异常和内存泄漏

- 3000以上并发，跑一次测试，不会有任何问题，当跑完一次测试之后不关闭程序，再跑一次，会在FastMM4.FastGetMem 中触发异常
![image](https://user-images.githubusercontent.com/3221597/52171901-39d4aa80-27a0-11e9-8491-6eb42818fb7e.png)


- 5000以上并发，程序退出时会在FastMM4.InsertMediumBlockIntoBin 中触发异常
![image](https://user-images.githubusercontent.com/3221597/52171895-fa0dc300-279f-11e9-98c6-ee9d07f56afa.png)

```
function TIocpCrossSocket._NewIoData: PPerIoData;
begin
  GetMem(Result, SizeOf(TPerIoData));
  FillChar(Result^, SizeOf(TPerIoData), 0);
end;

procedure TIocpCrossSocket._FreeIoData(P: PPerIoData);
begin
  if (P = nil) then Exit;

  P.CrossData := nil;
  P.Callback := nil;
  FreeMem(P, SizeOf(TPerIoData));
end;
```

**Comment by winddriver** [comment]
试了下 robertodellapasqua 提供的 IntelIPP_Delphi，编译通不过，pony5551 能否上传一份你测试的 intelTBB

**Comment by pony5551** [comment]
FastMM4-release_4992 也一样有问题啊！

**Comment by pony5551** [comment]
https://github.com/pony5551/PnHttpSysServer/tree/master/demos
我把intelTBB传到这里了

**Comment by pony5551** [comment]
测试了半天，感觉是interface的问题啊！

**Comment by winddriver** [comment]
的确是 interface 的问题，我尝试修改了下代码，不从连接队列中释放连接对象，测试过程中就不会出现异常了，测试了 FastMM4, IntelTBB, BrainMM, ScaleMM2, 都没有异常

但这并不能解决问题，这只是确定了异常是由于 interface 对象的释放引起的, 怎么处理这个奇怪的问题, 我暂时还没有头绪

```
procedure TAbstractCrossSocket.TriggerConnecting(AConnection: ICrossConnection);
begin
  AConnection.ConnectStatus := csConnecting;

  _LockConnections;
  try
    if not FConnections.ContainsKey(AConnection.UID) then
      Inc(FConnectionsCount);

    FConnections.AddOrSetValue(AConnection.UID, AConnection);
  finally
    _UnlockConnections;
  end;
end;

procedure TAbstractCrossSocket.TriggerDisconnected(AConnection: ICrossConnection);
begin
  AConnection.ConnectStatus := csClosed;

  _LockConnections;
  try
    if not FConnections.ContainsKey(AConnection.UID) then Exit;
// 不释放连接对象就不会出现异常
//    FConnections.Remove(AConnection.UID);
    Dec(FConnectionsCount);
  finally
    _UnlockConnections;
  end;

  LogicDisconnected(AConnection);

  if Assigned(FOnDisconnected) then
    FOnDisconnected(Self, AConnection);
end;
```

**Comment by pony5551** [comment]
这里释放了下面怎么还有两处调用啊？

**Comment by pony5551** [comment]
最好的方法是不用接口继承？这样性能可能会快一点！

**Comment by winddriver** [comment]
因为是接口，有引用计数保护的，从队列中移除并不表示就释放了，在引用计数减到 0 的时候才会触发释放，这也是我用接口的原因

**Comment by winddriver** [comment]
接口这点性能损失相较于它提供的便利完全可以忽略

**Comment by robertodellapasqua** [comment]
I get the AV in the procedure TIocpCrossSocket.Listen, on the loop _NewAccept
If can help, if I put a Sleep(1) inside the loop the AV disappear
indeed there's management of interfaces

another thing
on _Lock _Unlock triggers can you try to replace TMonitor with a lightweight TCriticalSection (if the queue management of TMonitor cause deadlocks somewhere)

**Comment by pony5551** [comment]
//在zs中使用CrossSocket，因为服务器会在一个实例中创建多个
  //如果给引用计数，在释放时会造成内存泄漏，所以这里我们使用zs的接口框架，不用引用计数
  //这里是我修改的，已通知过作者,by QQ600585
  TAbstractCrossSocket = class abstract(TCoreClassInterfacedObject, ICrossSocket)

**Comment by pony5551** [comment]
zs的群主改过一份

**Comment by pony5551** [comment]
> I get the AV in the procedure TIocpCrossSocket.Listen, on the loop _NewAccept
> If can help, if I put a Sleep(1) inside the loop the AV disappear
> indeed there's management of interfaces
> 
> another thing
> on _Lock _Unlock triggers can you try to replace TMonitor with a lightweight TCriticalSection (if the queue management of TMonitor cause deadlocks somewhere)

It's not what you said！

**Comment by robertodellapasqua** [comment]
I get systematically the AV in the _newaccept loop function, and if fail to acceptex then debug mode do a log and the malloc of getUnicodeString in system get AV in malloc too. All scrambled.

**Comment by pony5551** [comment]
TCoreClassInterfacedObject = class(TInterfacedObject)
  protected
    function _AddRef: longint; {$IFNDEF WINDOWS} cdecl {$ELSE} stdcall {$ENDIF};
    function _Release: longint; {$IFNDEF WINDOWS} cdecl {$ELSE} stdcall {$ENDIF};
  public
    procedure AfterConstruction; override;
    procedure BeforeDestruction; override;
  end;


{ TCoreClassInterfacedObject }

function TCoreClassInterfacedObject._AddRef: Integer;
begin
  Result := 1;
end;

function TCoreClassInterfacedObject._Release: Integer;
begin
  Result := 1;
end;

procedure TCoreClassInterfacedObject.AfterConstruction;
begin
end;

procedure TCoreClassInterfacedObject.BeforeDestruction;
begin
end;

然后替换所有 TInterfacedObject为TCoreClassInterfacedObject

**Comment by pony5551** [comment]
好像还有别的地方，我找一下

**Comment by robertodellapasqua** [comment]
![clipboard01](https://user-images.githubusercontent.com/24314583/52175386-52928f80-27a3-11e9-93a1-43b796bae0cf.jpg)

{$IFDEF MSWINDOWS}
function SysErrorMessage(ErrorCode: Cardinal; AModuleHandle: THandle): string;
var
  Buffer: PChar;
  Len: Integer;
  Flags: DWORD;
begin
  Flags := FORMAT_MESSAGE_FROM_SYSTEM or
    FORMAT_MESSAGE_IGNORE_INSERTS or
    FORMAT_MESSAGE_ARGUMENT_ARRAY or
    FORMAT_MESSAGE_ALLOCATE_BUFFER;

  if AModuleHandle <> 0 then
    Flags := Flags or FORMAT_MESSAGE_FROM_HMODULE;

  { Obtain the formatted message for the given Win32 ErrorCode
    Let the OS initialize the Buffer variable. Need to LocalFree it afterward.
  }
  Len := FormatMessage(Flags, Pointer(AModuleHandle), ErrorCode, 0, @Buffer, 0, nil);

  try
    { Remove the undesired line breaks and '.' char }
    while (Len > 0) and ((Buffer[Len - 1] <= #32) or (Buffer[Len - 1] = '.')) do
      Dec(Len);
    { Convert to Delphi string }
    SetString(Result, Buffer, Len);
  finally
    { Free the OS allocated memory block }
    **

> LocalFree(HLOCAL(Buffer)); //// AV

**
  end;
end;

00007FFA1FB1D25D 7409             jz $00007ffa1fb1d268
00007FFA1FB1D25F 498B00           mov rax,[r8] //// bad pointer
00007FFA1FB1D262 48894258         mov [rdx+$58],rax
00007FFA1FB1D266 EB25             jmp $00007ffa1fb1d28d
00007FFA1FB1D268 4C8B4250         mov r8,[rdx+$50]
00007FFA1FB1D26C 4D85C0           test r8,r8
00007FFA1FB1D26F 7452             jz $00007ffa1fb1d2c3

**Comment by robertodellapasqua** [comment]
btw. it is a good practice to do this in the MM replace unit:

var
  OldMemoryManager: TMemoryManagerEx;

initialization
GetMemoryManager(OldMemoryManager);
SetMemoryManager(SEAMemoryManager);

finalization
SetMemoryManager(OldMemoryManager);

because Delphi with some MM produces exceptions on exit of the app, probably calling freemem after MM disposition; with this simple trick the app correctly exit

**Comment by robertodellapasqua** [comment]
*offtopic* Pony555 a question, do you have modified mormot to use WSA_OVERLAPPED I/O instead of serial (blocking threads/one thread for one connection)?

**Comment by pony5551** [comment]
> _offtopic_ Pony555 a question, do you have modified mormot to use WSA_OVERLAPPED I/O instead of serial (blocking threads/one thread for one connection)?
Yes, I use HTTP. Sys combined with iocp!

**Comment by robertodellapasqua** [comment]
great! I love classic OOP with manual mem management, create, use, free, no GC as Java, nor ARC, neither automatic reference counting :-P
do your interface overriding methods have solved the MM troubles?

**Comment by pony5551** [comment]
Need to manage their own objects! My code doesn't solve the problem!

**Comment by robertodellapasqua** [comment]
Indeed interfaces have many advantages, for example when you can define a class derived from many interfaces, or if you want move and call objects easily in external libraries without rtti etc....
anyway, we need figure out the problem... because this cross socket server is really a great code!

**Comment by robertodellapasqua** [comment]
> 试了下 robertodellapasqua 提供的 IntelIPP_Delphi，编译通不过，pony5551 能否上传一份你测试的 intelTBB

use RDPMM64 as first unit, output the exe in the same folder with the two SEA dlls

**Comment by robertodellapasqua** [comment]
ok, I use MS heap, from Windows 2016 server it's very fast, very scalable and low fragmentation.

Attached my unit where I use Fillchar, Move and Pos from Intel IPP.

A benchmark with FastMM4:
root@roberto-virtual-machine:/home/roberto# ab -n 10000 -c 1000 -k -r 
Concurrency Level:      1000
Time taken for tests:   5.953 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    10000
Total transferred:      24770000 bytes
HTML transferred:       22480000 bytes
**Requests per second:    1679.92 [#/sec] (mean)**
Time per request:       595.267 [ms] (mean)
Time per request:       0.595 [ms] (mean, across all concurrent requests)
Transfer rate:          4063.63 [Kbytes/sec] received

Same with MsHeap (Windows 2016) and Intel IPP SIMD FillChar,Move
Concurrency Level:      1000
Time taken for tests:   5.620 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    10000
Total transferred:      24770000 bytes
HTML transferred:       22480000 bytes
**Requests per second:    1779.26 [#/sec] (mean)**
Time per request:       562.033 [ms] (mean)
Time per request:       0.562 [ms] (mean, across all concurrent requests)
Transfer rate:          4303.92 [Kbytes/sec] received

[RDPHeap64.pas.txt](https://github.com/winddriver/Delphi-Cross-Socket/files/2825697/RDPHeap64.pas.txt)

So for now, with this I get no errors, if meantime somebody can discover the TBB and other troubles please let me know.
CIAO

**Comment by robertodellapasqua** [comment]
ok, the error still exist, look if I disable the SERIALIZATION on the MS HEAP, permitting simultaneous threads to read/write to the heap BOOM! There are double free somewhere or null pointers.
(See the 8 I/O threads doing something of wrong)

If I do a new Delphi APP works perfectly also with NO_SERIALIZATION.

![clipboard02](https://user-images.githubusercontent.com/24314583/52180531-a9ba5380-27e7-11e9-8959-ebb133151046.jpg)

**Comment by robertodellapasqua** [comment]
please, download https://www.nexusdb.com/support/index.php?q=node/27156 and run CodeWatch to find and fix the error

**Comment by pony5551** [comment]
昨天在群里请教大佬，大佬分析出了一个问题！
好像说把[weak]改成[unsafe]会好，不过还有个问题就是退出时会抛异常！
太晚了，明天继续

**Comment by robertodellapasqua** [comment]
well, unsafe will be faster with a lot of objects
for example if you have a tdictionary, you have to free manually the object
this sound ok!

or you can follow this link
http://blog.synopse.info/post/2012/06/18/Circular-reference-and-zeroing-weak-pointers
and this
https://www.codeproject.com/Articles/1252175/Fixing-Delphis-Interface-Limitations

I suggest avoid weak reference, because they are managed with a hash table and a global lock
let's manage the memory directly in a server without arc overhead!

**Comment by pony5551** [comment]
![qq 20190204085410](https://user-images.githubusercontent.com/36391163/52185430-94324180-285a-11e9-8ff2-bc09076917c1.png)
这里AItem是无效的野指针

**Comment by robertodellapasqua** [comment]
![2](https://user-images.githubusercontent.com/24314583/52258993-6106b580-2921-11e9-9c74-ca7de182b7a5.png)

ProcessIO is useless (there is the time waiting on the iocomp)
but the other first routines can be optimized

**Comment by robertodellapasqua** [comment]
![1](https://user-images.githubusercontent.com/24314583/52259727-07ec5100-2924-11e9-9b96-3b839efc59f7.png)

for example you can replace TFileStream with **TBufferedStream**, we should see a negligible gain in throughput

https://delphiaball.co.uk/2016/04/29/faster-filestream-tbufferedfilestream/

(or I can provide you a cache I did years ago for faster retrieve of frequent files in from memory)(anyway the kernel file cache should be enough)

Now I have no more time, I need proceed with work, so please, if you can take full version of NexusQA, trap the memory overrun, make the interfaces unsafe with explicit free(), don't inherit twebsocket from httpconnection, make them separated: when a httpconnection get 101 upgrade and start the websocket, then free the httpconnection to save memory and resource not more needed.

With those corrections this library can become a new standard in the delphi community because is wonderful.

Best.

**Comment by robertodellapasqua** [comment]
how are you?
Look, under stress test, both FastMM4-avx (the latest), TCmalloc, TBB and MsHeap are producing artifacts, odd exception pointers in system units, failed requests. Only without replacing MM works without errors. That's odd situation. I don't know if the problem is the compiler or the code.

Btw. in httpserver if you want to renounce to some internal checking (if fileexists, if characters are permitted) replace

LStream := TFile.Open(AFileName, TFileMode.fmOpen, TFileAccess.faRead, TFileShare.fsReadWrite);
with
LStream:= TFileStream.Create(AFileName, fmOpenRead or fmShareDenyNone);

from 1360 req/sec to 1880 req/sec

What's your opinion, is TFileOpen parameters checking useful?
Somebody has Nexus codewatch license to catch the double free and pointers problems with other MM?

**Comment by pony5551** [comment]
我们中国在过新年！祝大家新年快乐，财源滚滚来！

**Comment by robertodellapasqua** [comment]
let's check using delphi 10.3, because with this latest version, using default mm, benchmarks show the server unreliable (sometimes failed requests), so the problem should be somewhere in the code.

**Comment by robertodellapasqua** [comment]
10.3 situation compiling crosshttpconsole, without any change:
-> linux, works without fails, very fast asking http//IP/hello; if you ask a files not found the http returns a string "Not found", but fails a lot of requests
-> win64, the exe when start throw this exception and exit (default MM):
C:\Exes>crosshttpconsole

Windows Server 2016 (Version 10.0, Build 14393, 64-bit Edition)
EAccessViolation: Access violation at address 000000000044155A in module 'CrossHttpConsole.exe'. Read of address 0000000000000040
Unexpected Memory Leak
An unexpected memory leak has occurred. The unexpected small block leaks are:

9 - 24 bytes: TDictionary<System.UInt64,Net.CrossSocket.Base.ICrossListen>.TValueCollection x 1, TDictionary<System.UInt64,Net.CrossSocket.Base.ICrossConnection>.TValueCollection x 1, TObject x 2
57 - 72 bytes: Unknown x 2
89 - 104 bytes: TDictionary<System.UInt64,Net.CrossSocket.Base.ICrossConnection> x 1, TDictionary<System.UInt64,Net.CrossSocket.Base.ICrossListen> x 1
521 - 568 bytes: TCrossHttpServer x 1

**Comment by robertodellapasqua** [comment]
Allen kindly wrote me:
With IOCP the reference to the interface is not maintained when it is passed as a pointer to the kernel, so it is up to the developer to make sure you manually add a reference everywhere this could happen.  Delphi automatically attempts to destroy it when it thinks it is no longer required, when it reality the pointer is waiting to return from the kernel.  The timing of this is impacted by how quickly the completion happens.  This can lead to some really hard to debug/identity memory management issues.

**Comment by pony5551** [comment]
按楼上老兄说的只能不使用引用计数，然后自己维护对象了

**Comment by pony5551** [comment]
发现个问题，在传递给io之前使用
obj._AddRef;
这样就不会有释放问题了！不过会有其他问题

TIocpCrossSocket._NewReadZero.WSARecv : System Error.  Code: 10053(2745), 你的主机中的软件中止了一个已建立的连接。
TIocpCrossSocket._NewReadZero.WSARecv : System Error.  Code: 10053(2745), 你的主机中的软件中止了一个已建立的连接。

**Comment by robertodellapasqua** [comment]
as far as I read in the Delphi docs, you should carefully manage the _AddRef and _Release pair for istantiate/deistantiate every interface, do we know when call the _Release, when the object is not more used, out of scope?

**Comment by pony5551** [comment]
可以在需要的时候调用Release,不过好像我没调用Release也没发现问题！
你可以试试

**Comment by pony5551** [comment]
我到现在都不知道什么原因，接口会释放两次！

**Comment by robertodellapasqua** [comment]
have you try to use [weak], then use _Release() to free the interface?

**Comment by pony5551** [comment]
没试过，好像也是可以的！！！

**Comment by pony5551** [comment]
等待Souledge更新项目，我都有些焦虑了！

**Comment by robertodellapasqua** [comment]
do you have solved without mem leaks or double free twice?

**Comment by pony5551** [comment]
[unsafe]或者obj._Addref;都可以解决！
我个人感觉在AcceptEx,WSARecv,WSASend前使用obj._Addref;比较好

**Comment by robertodellapasqua** [comment]
using [unsafe] the app will be more scalable, because we avoid to use the global hash collectio, bu n, with single locking, of managed mode), but should free it manually after the use

**Comment by pony5551** [comment]
我修改的文件在这里
https://github.com/pony5551/Delphi-Cross-Socket/blob/master/Net/Net.CrossSocket.Iocp.pas
你可以测试！

**Comment by robertodellapasqua** [comment]
thank you very much
I get no leaks nor exceptions both with default mm or ms heap
still a minor trouble with IntelTBB, the first time I run the benchmark it freeze at the last packet,
relaunching the bench works fine
(when I have time will modify using unsafe attribute)

**Comment by robertodellapasqua** [comment]
ok, look, only if I use TBB Intel I get this error onstart:
Windows Server 2016 (Version 10.0, Build 14393, 64-bit Edition)
EStackOverflow: Stack overflow
Unexpected Memory Leak
An unexpected memory leak has occurred. The unexpected small block leaks are:

57 - 72 bytes: Unknown x 12
all other leaks are solved

**Comment by robertodellapasqua** [comment]
ok, try to run with IBB, seems better to detect leaks, I get in two points:
TPerlRegEx (seems a leak into this system unit)
and in Net.Cross.IoCp line 671   _FreeIoData(LPerIoData);

**Comment by winddriver** [comment]
手动加 _AddRef 和 _Release 这纯属多余，接口的引用计数机制本身就能确保在将一个接口赋值给另一个接口变量时，接口引用计数自动加1，将接口变量赋值为 nil 时，接口引用计数自动减1

pony5551 的修改我看过也测试过了，那几个 _AddRef 只是额外的把引用计数又多加了一次，在给 LPerIoData.CrossData 赋值时那几个 _AddRef 能对应上 _FreeIoData 中的 _Release，所以这几个不会有什么问题，当然也不会有任何帮助；但是 _HandleRead 中 LConnection._AddRef 就有问题了，这额外的引用计数增加并没有对应的 _Release，就造成这个连接接口永远不可能被释放，必然引起大量内存泄露

现在我不好说到底问题出在哪里，因为使用 Delphi 自带的内存管理器，不会有任何问题，只有使用第三方的内存管理器才会出现问题，估计需要花时间去读 Delphi 接口管理相关的代码才能找出问题所在了

**Comment by robertodellapasqua** [comment]
so the problem seems to happen when we call an async API that returns immediately, Delphi dispose the interface, and meantime the kernel has not finished to return the key, on the callback will try to use a RecordInvoke for example on a disposed structure; or maybe mm delphi management uses async temporizer/threads producing troubles with third-party managers; well meantime thanks for your info

**Comment by pony5551** [comment]
我个人强烈建议：取消部分接口设计，自己管理object要不然很容易爆雷

**Comment by robertodellapasqua** [comment]
for curiosity apache bench on i7 4 core:
standard mm -> 17000 op/s
ms heap + subset IPP -> 35000 op/s (double throughput, let's imagine in a high core processor)
(if can be useful also the heap manager of the OS can intercept mem leaks and errors, use the flag HEAP_GENERATE_EXCEPTIONS, if enabled will produce the similar info of TBB with exceptions on mem leaks)

**Comment by robertodellapasqua** [comment]
this is very nice:

windows 2016 server
**default mm**

Server Software:        CrossHttpServer/2.0
Server Hostname:        192.168.1.166
Server Port:            8000
Document Path:          /hello
Document Length:        11 bytes
Concurrency Level:      100
Time taken for tests:   4.020 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
Total transferred:      14200000 bytes
HTML transferred:       1100000 bytes
**Requests per second:    24872.78 [#/sec] (mean)**
Time per request:       4.020 [ms] (mean)
Time per request:       0.040 [ms] (mean, across all concurrent requests)
Transfer rate:          3449.15 [Kbytes/sec] received

**ms heap + ipp 360% faster (!!!)**
Server Software:        CrossHttpServer/2.0
Server Hostname:        192.168.1.166
Server Port:            8000
Document Path:          /hello
Document Length:        11 bytes
Concurrency Level:      100
Time taken for tests:   1.133 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
Total transferred:      14200000 bytes
HTML transferred:       1100000 bytes
**Requests per second:    88227.92 [#/sec] (mean)**
Time per request:       1.133 [ms] (mean)
Time per request:       0.011 [ms] (mean, across all concurrent requests)
Transfer rate:          12234.73 [Kbytes/sec] received

But under linux, under identical configuration, I get 
**Requests per second:    2247.26 [#/sec] (mean)**

So IMHO windows code is excellent, but epoll linux should be rewritten without the global locking I/O queue
Pony555 let me know if you can figure out the last little memory leak found in the windows version
thanks

**Comment by pony5551** [comment]
你可以看看这里的处理方式
https://github.com/PassByYou888/ZServer4D

**Comment by pony5551** [comment]
能不能把你的ms heap + ipp 
代码传上来？

**Comment by robertodellapasqua** [comment]
[MSHeap_IPP.zip](https://github.com/winddriver/Delphi-Cross-Socket/files/2871822/MSHeap_IPP.zip)
幹得好
btw. what is ZServer4D? A Rest server?

**Comment by pony5551** [comment]
那个项目上有说明，是个很不错的云服务器架构

**Comment by robertodellapasqua** [comment]
a curiosity, why you china men are using avatars to publish source code in github, without put a link to linkedin page, or webpage with real data, email, etc. ? :-)

**Comment by pony5551** [comment]
这个我也不知道！
我们中国一个delphi讨论板，http://bbs.2ccc.com

**Comment by robertodellapasqua** [comment]
I have enabled patch SIMD for Pos() functions, please test it
on I7 quadcore jump from 27K req/s to 92K req/s (believe!)

Server Software:        CrossHttpServer/2.0
Server Hostname:        192.168.1.166
Server Port:            8000
Document Path:          /hello
Document Length:        11 bytes
Concurrency Level:      100
Time taken for tests:   3.703 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
Total transferred:      14200000 bytes
HTML transferred:       1100000 bytes
Requests per second:    **27002.22 [#/sec] (mean)**
Time per request:       3.703 [ms] (mean)
Time per request:       0.037 [ms] (mean, across all concurrent requests)
Transfer rate:          3744.45 [Kbytes/sec] received

Server Software:        CrossHttpServer/2.0
Server Hostname:        192.168.1.166
Server Port:            8000
Document Path:          /hello
Document Length:        11 bytes
Concurrency Level:      100
Time taken for tests:   1.094 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
Total transferred:      14200000 bytes
HTML transferred:       1100000 bytes
Requests per second:    **91442.20 [#/sec] (mean)**
Time per request:       1.094 [ms] (mean)
Time per request:       0.011 [ms] (mean, across all concurrent requests)
Transfer rate:          12680.46 [Kbytes/sec] received

If you use my patches please put a link to my website www.dellapasqua.com and maybe forward me some jobs internet related, fullstack, cloud, embedded, why not? Regards
[RDPRTL.zip](https://github.com/winddriver/Delphi-Cross-Socket/files/2872302/RDPRTL.zip)
[TestPOs.zip](https://github.com/winddriver/Delphi-Cross-Socket/files/2872303/TestPOs.zip)

**Comment by pony5551** [comment]
Please try the new version of fastmm4

**Comment by robertodellapasqua** [comment]
I have try the new FastMM4-avx with AVX all enabled, and the results are slower than the default MM of RIO 10.3.1

**Comment by robertodellapasqua** [comment]
you are right!
using FastMM4-avx, disabling all the debug, leak test, and enabling the 3 flags for multithreading I get the same performances of RDPRTL
the problem is that I have read in the forums that fastmm4-avx has troubles not yet solved

**Comment by winddriver** [comment]
问题已经完美解决，感谢 robertodellapasqua 和 pony5551 这段时间的辛苦付出

**Comment by robertodellapasqua** [comment]
Souledge has corrected the sources, updated with TBB:
Server Software:        CrossHttpServer/2.0
Server Hostname:        192.168.1.166
Server Port:            8000
Document Path:          /hello
Document Length:        11 bytes
Concurrency Level:      100
Time taken for tests:   1.015 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
Total transferred:      14200000 bytes
HTML transferred:       1100000 bytes
Requests per second:    **98514.50** [#/sec] (mean)
Time per request:       1.015 [ms] (mean)
Time per request:       0.010 [ms] (mean, across all concurrent requests)
Transfer rate:          13661.19 [Kbytes/sec] received

So:
default 27K/s
win2016 heap 91K/s
**intel TBB 98K/s**
[RDP_IntelMMRTL.zip](https://github.com/winddriver/Delphi-Cross-Socket/files/2872759/RDP_IntelMMRTL.zip)
But RDPMM64 as first unit in project source

**Comment by robertodellapasqua** [comment]
> 问题已经完美解决，感谢 robertodellapasqua 和 pony5551 这段时间的辛苦付出

thank you very much for this excellent library!

**Comment by robertodellapasqua** [comment]
please rem the Pos() patches, because I have found a bug, will correct it soon

**Comment by pony5551** [comment]
> please rem the Pos() patches, because I have found a bug, will correct it soon

Can your mm be cross-platform

**Comment by robertodellapasqua** [comment]
yes, I have Linux version too, I'm correcting some bugs of the RTL, need do workarounds.

**Comment by pony5551** [comment]
期待你的作品！

**Comment by robertodellapasqua** [comment]
I'm facing problems with CompareText(), because the IPP version behaves differently than Delphi RTL (some third-party components can't work); IMHO tomorrow will upload the version with Pos() corrected

**Comment by robertodellapasqua** [comment]
Pony do you agree to rewrite epoll layer? Do you help me?

**Comment by robertodellapasqua** [comment]
Next patch:
Under Windows we should replace TFile stream with NT TransmitFile() API overlapped, this use the kernel cache. We will have 10x speedup on file serving

**Comment by pony5551** [comment]
> Pony do you agree to rewrite epoll layer? Do you help me?
> Sorry can I permit, does Pony555, Souledge have a company webpage? Or are you graduate students? Where are you located? If you wish can write me privately, also I have friends in Hong Kong :-)

What's wrong with epoll?

**Comment by pony5551** [comment]
> Next patch:
> Under Windows we should replace TFile stream with NT TransmitFile() API overlapped, this use the kernel cache. We will have 10x speedup on file serving

Asynchronous file manipulation is also possible under Linux！

**Comment by pony5551** [comment]
> Souledge has corrected the sources, updated with TBB:
> Server Software: CrossHttpServer/2.0
> Server Hostname: 192.168.1.166
> Server Port: 8000
> Document Path: /hello
> Document Length: 11 bytes
> Concurrency Level: 100
> Time taken for tests: 1.015 seconds
> Complete requests: 100000
> Failed requests: 0
> Keep-Alive requests: 100000
> Total transferred: 14200000 bytes
> HTML transferred: 1100000 bytes
> Requests per second: **98514.50** [#/sec] (mean)
> Time per request: 1.015 [ms] (mean)
> Time per request: 0.010 [ms] (mean, across all concurrent requests)
> Transfer rate: 13661.19 [Kbytes/sec] received
> 
> So:
> default 27K/s
> win2016 heap 91K/s
> **intel TBB 98K/s**
> [RDP_IntelMMRTL.zip](https://github.com/winddriver/Delphi-Cross-Socket/files/2872759/RDP_IntelMMRTL.zip)
> But RDPMM64 as first unit in project source

你这个在我的机器上效果不是很好！和默认的差不多，最多提高10%-20%

**Comment by robertodellapasqua** [comment]
I did the test in a VMware windows 2016 server, 4 cpu, and this apachebench:
ab -n 10000 -c 100 -k -r http://192.168.1.124:8000/hello

**Comment by pony5551** [comment]
我的是4核心虚拟机，win10
i7 6820hq

ab -n 100000 -c 1000 -k -r http://localhost:8000/hello


我也用windows server 2019测试了，比win10好一点而已！

能不能上传你的测试代码

**Comment by robertodellapasqua** [comment]
https://docs.microsoft.com/it-it/windows/desktop/api/mswsock/nf-mswsock-transmitfile

**TF_USE_KERNEL_APC** and **TF_WRITE_BEHIND** will make sendfile completely asynchronous with a callback when completed or failed (using overlapped IOCP binding),
is the same way used in IIS https://docs.microsoft.com/en-us/previous-versions/iis/6.0-sdk/ms525816%28v%3dvs.90%29

before call transmitfile use this flag when open the file handle A handle to the open file that the 
**TransmitFile** function transmits. Since the operating system reads the file data sequentially, you can improve caching performance by opening the handle with **FILE_FLAG_SEQUENTIAL_SCAN**

**Comment by robertodellapasqua** [comment]
> 我的是4核心虚拟机，win10
> i7 6820hq
> 
> ab -n 100000 -c 1000 -k -r http://localhost:8000/hello
> 
> 我也用windows server 2019测试了，比win10好一点而已！
> 
> 能不能上传你的测试代码

try to use -100 concurrency instead of -1000, because is the client ab.exe that is not able to support highest speed with too much concurrent items

**Comment by robertodellapasqua** [comment]
of course to do a real test scenario should use a lan, with many clients, and with tools to simulate web calls to many path/files/rest

**Comment by robertodellapasqua** [comment]
with epoll calling files I have notices under stress test many failed requests, and poor throughput; I did a read of the source, and probably the bottleneck is the use of a single queue container with locking semaphores for all the sockets, we can "multiplex" or free from this

**Comment by robertodellapasqua** [comment]
https://www.winsocketdotnetworkprogramming.com/winsock2programming/winsock2advancedscalableapp6.html

**Comment by pony5551** [comment]
> with epoll calling files I have notices under stress test many failed requests, and poor throughput; I did a read of the source, and probably the bottleneck is the use of a single queue container with locking semaphores for all the sockets, we can "multiplex" or free from this

你这个提议之前有人提出过，每一个线程一个独立的listen,但是Souledge 的时间太少了！
你可以修改一个分支然后告诉我们，Souledge 看到了可以参考代码，这样就节省了很多时间！
好像ioco也可以这样做，或者用windows rio(Registered I/O)

**Comment by pony5551** [comment]
-100 也测试过了，效果只是提高一点点

**Comment by pony5551** [comment]
你有什么即时聊天工具吗？我们可以用即时聊天工具交流！
比如我们的www.qq.com

**Comment by robertodellapasqua** [comment]
I use this command ab -n 10000 -c 100 -k -r http://192.168.1.124:8000/hello
if you get same numbers probably have not put the RDPMM64 as first unit in the project source clause
do you have skype? else I install qq

**Comment by robertodellapasqua** [comment]
now I'm very busy with job things, I'll come back within some days
btw. IMHO Souledge can modify $IFDEF USEKERNEL TransmitFile $ENDIF, I open a feature request

**Comment by pony5551** [comment]
ab -n 10000 -c 100 -k -r http://localhost:8000/hello

我就是这样用的

**Comment by pony5551** [comment]
Skype charges for many features!

Qq all communication functions are free including video call！

My qq number is 7180001

**Comment by robertodellapasqua** [comment]
I catch you in the next days, thank you

**Comment by robertodellapasqua** [comment]
> > with epoll calling files I have notices under stress test many failed requests, and poor throughput; I did a read of the source, and probably the bottleneck is the use of a single queue container with locking semaphores for all the sockets, we can "multiplex" or free from this
> 
> 你这个提议之前有人提出过，每一个线程一个独立的listen,但是Souledge 的时间太少了！
> 你可以修改一个分支然后告诉我们，Souledge 看到了可以参考代码，这样就节省了很多时间！
> 好像ioco也可以这样做，或者用windows rio(Registered I/O)

rio is useful only in udp architecture, is not for tcp
http://www.serverframework.com/asynchronousevents/2012/08/winsock-registered-io-io-completion-port-performance.html
how you do async file transfer in Linux? The kernel doesn't support it, neither using libaio. Nginx uses a thread pool, one thread is used coupled with sendfile()
https://www.nginx.com/blog/thread-pools-boost-performance-9x/

**Comment by pony5551** [comment]
![qq 20190218181552](https://user-images.githubusercontent.com/36391163/52943956-4b759f00-33a9-11e9-8a8b-f16afa39b6d4.png)
这里是不是要调整一下

**Comment by robertodellapasqua** [comment]
do you know a web or socket china proxy (without vpn) to permits to use QQ Windows client from Europe?
I can use QQ only from Iphone, and this is not good

**Comment by robertodellapasqua** [comment]
my QQ id is BOB74 account number 3470314349
